src/turbolinks/visit.coffee

```diff
@@ -56,11 +56,12 @@ class Turbolinks.Visit
   loadCachedSnapshot: ->
     if snapshot = @getCachedSnapshot()
       isPreview = @shouldIssueRequest()
-      @render ->
-        @cacheSnapshot()
-        @controller.render({snapshot, isPreview}, @performScroll)
-        @adapter.visitRendered?(this)
-        @complete() unless isPreview
+      if snapshot.isPreviewable() or not isPreview
```

Comment:
I think this check should be moved to `getCachedSnapshot`. `getCachedSnapshot` should return nothing if a usable snapshot can't be found. Just like it does for snapshots that lack a requested anchor.

```coffee
getCachedSnapshot: ->
  if snapshot = @controller.getCachedSnapshotForLocation(@location)
    if not @location.anchor? or snapshot.hasAnchor(@location.anchor)
      if @action is "restore" or snapshot.isPreviewable()
        snapshot
```

This way we won't need to change `loadCachedSnapshot` or `hasCachedSnapshot`.